manifests/vrrp/instance.pp

```diff
@@ -167,6 +177,7 @@
   $virtual_ipaddress_int                                            = undef,
   $virtual_ipaddress_excluded                                       = undef,
   $virtual_routes                                                   = undef,
+  $virtual_rules                                                    = undef,
```

Comment:
Can you please add a Datatype for this parameter that enforces that hash structure? It might be a good idea to enforce that it's always an array of hashes. An array of structs might be the best approach here.

Reply:
Somethig like this?:
`Optional[Array[Hash]] $virtual_rules = undef`

Reply:
A struct would allow you to enforce the name of the keys within the hash and also having specific datatypes per key. The documentation is here:
https://puppet.com/docs/puppet/latest/lang_data_abstract.html#the-struct-data-type
Our rsyslog module already has a bunch of those: https://github.com/voxpupuli/puppet-rsyslog/blob/master/types/actions.pp

Reply:
Thanks for the examples. After templating this (and before your examples)...:
```
Optional[Array[Struct[
  Optional[from]  => String[1],
  Optional[to]    => String[1],
  Optional[dev]   => String[1],
  Optional[table] => String[1]]] => undef,
```
.. I did choose the `Optional[Array[Hash]]` variant.

Reply:
Updated to arrays of Structs =)